### PR TITLE
JIT: CPU simulator

### PR DESCRIPTION
**Pull request by asi1024**
TODO:
- [ ] builtin functions
- [ ] device function call
- [ ] where inputs share the same memory
- [ ] tests

```py
import numpy
import cupy
from cupyx import jit


@jit.rawkernel()
def f(x, y):
    tid = jit.threadIdx.x + jit.blockDim.x * jit.blockIdx.x
    y[tid] = x[tid]


x = numpy.random.randint(100, size=30)
y = numpy.random.randint(100, size=30)
f.simulate((5,), (6,), (x, y))
assert bool((x == y).all())
```

### CODE REVIEW

**Comment by kmaehashi** [comment]
Discussed offline: Launching host threads of the same number as CUDA threads could be a problematic. How about running them sequentially and use yield when syncthread is needed?

File: cupyx/jit/_simulate.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+def simulate(func, args, grid_dim, block_dim):
+    new_args = []
+    for x in args:
+        if isinstance(x, (numpy.ndarray, cupy.ndarray)):
+            read = numpy.zeros(x.shape, dtype=numpy.bool_)
+            write = numpy.zeros(x.shape, dtype=numpy.bool_)
+            x = _Global(x, read, write)
+        new_args.append(x)
+    n_workers = numpy.prod(grid_dim) * numpy.prod(block_dim)
+    with concurrent.futures.ThreadPoolExecutor(n_workers) as executor:
```
Comment (emcastillo) [inline cupyx/jit/_simulate.py:75]: GIL is going to make this sequential anyway, so how about just looping and not using threads or futures at all?

Also, utilities for executing a single thread, or a single block to check indexes for arrays access could be handy here.

**Comment by asi1024** [comment]
> Discussed offline: Launching host threads of the same number as CUDA threads could be a problematic. How about running them sequentially and use yield when syncthread is needed?

@kmaehashi @emcastillo 
I've thought about it after the discussion, and it seems that this idea will require changing the syntax to `yield from jit.syncthreads()`.
Now, I want to change the syntax to simply write `yield "<label>"` as a syncthreads, and deprecate `jit.synchtreads`. How do you think about it?

**Comment by takagi** [comment]
Is it too much to implement another naive compiler that transpiles CuPy JIT to Python and evaluate its generated result with `exec` so that we can have arbitrary semantics on `jit.syncthreads()`?

**Comment by kmaehashi** [comment]
Yeah @emcastillo and I were thinking of the similar idea, e.g., parse the user function and rewrite `syncthreads` etc to `await` or `yield` and execute it.

**Comment by emcastillo** [comment]
The thing is like we have to take into account block & grid sync (cooperative groups?).
Probably mem fences are Ok since we don't have real parallel execution but the barriers will force us to have live "threads" holding their own local contexts.
We discussed this and maybe `asyncio` is the cleanest way to go. Barriers are not directly available but they can be easily implemented.

**Comment by asi1024** [comment]
How about simulating on AST? For example, when using warp shuffle operations, we want to synchronize the threads for each operation, but I think that multi-threading or yield cannot deal with it well.

**Comment by emcastillo** [comment]
might be a good idea

**Comment by kmaehashi** [comment]
Offline discussion:
We would like to avoid both:
* launching thousands of host threads
* reinventing an interpreter that executes AST
So rewriting the target function with `asyncio` sounds the way to go. This should be doable by  by rewriting the target function `def` to `async def`, and rewrite `syncthreads` / ooperative groups / etc. with `await from`, all via AST.

**Comment by asi1024** [comment]
@kmaehashi I've considered how to implement it, but I think it would be difficult to support a CPU simulator without reinventing an interpreter that executes AST while ensuring perfect synchronization for identical warps.
Closing this PR until we find a suitable implementation policy.

**Pull request closed without merging**